The last bits of <errno.h> work

le-jzr opened the pull request:
TODO: more commentary
Not going to merge this at least until next week.

Basically just the conclusion of my past two weeks. Haven't tested coastline with it yet.
Note: Extra error codes from libposix are moved to <abi/errno.in> (with a fat comment explaining they are just there for convenience). It's the simplest solution, but I have a feeling others might disagree with it. Discuss in comments.

le-jzr commented:
Here goes more commentary:

First of all, I add `<_bits/opaque_handle.h>` which adds macro `opaque_handle(type_name)`, which expands to a typedef of a pointer to incomplete struct type. I plan to use this in the future to define types such as VFS file handles and the like.

`opaque_handle()` is conditionally (`#ifdef __OPAQUE_ERRNO__`) used to define `errno_t` as a pointer type. This allows us to prevent mixing of error codes and non-error values on a type level (see [my blog post](https://le-jzr.github.io/helenos/c/errno/2017/12/11/abusing-type-checking-for-fun-and-profit.html)).

In kernel, most syscalls return error number, but cannot be declared as returning `errno_t` because they are invoked indirectly from the syscall handler. For this reason `sys_errno_t` is additionally defined in the kernel as a type that holds error codes, but is represented as `sysarg_t`. There is no such type in `uspace`.

Pointers cannot be used by a `switch`, so as a workaround, there's additionally a `case_errno_t` (horrible name, I plan to either change it or remove it altogether, see below), defined as an alias for `long`, which is used for casting error codes in `switch` statements. Using `tools/srepl '(case_errno_t) ' ''`, it's deliberately easy to remove from the codebase.

Additionally, a lot of binaries return `errno` code from `main()`, but at the same time also return arbitrary magic constants (-1 and small positive integers). Instead of trying to resolve this situation, I introduced a nonstandard macro `EXIT_RC(rc)` (currently defined in `<stdlib.h>` alongside `EXIT_SUCCESS` and `EXIT_FAILURE`). This macro can be changed in the future to map the codes into a reserved range, or something similar. Currently it's either a simple cast or nop.

Finally, I reassigned error code numbers to positive values, and modified libposix to use libc's `errno.h`.

Planned but uncommited: remove `-D__OPAQUE_ERRNO__` from the default build, possibly remove the "scaffolding" as well (see below). 

Before merging, there are following unresolved questions:
- Does anyone mind having compatibility-only error constant names in core headers? It simplifies things considerably. See the comments in final `abi/include/abi/errno.in`.
- Should I keep the conditional `errno_t` definition, `case_errno_t` and `EXIT_RC()` until we get a custom checking tool, or should I clean them out before merging into mainline?
- If `case_errno_t` is kept, what form should it take? The current one is very unsemantic.

le-jzr commented:
Note: currently doesn't build or anything but `amd64` and untested with coastline. I'll get to it after xmas.

jxsvoboda commented:
You shouldn't use double leading underscore in macro names / symbolic variables. That is the part of the namespace reserved for the compiler. Just use a single leading underscore (e.g. _errno_xx)

If you are putting all the Exxx constants in a single header split into sections, you might want to make sure that code that is not supposed to make use of a particular section cannot do so by mistake (e.g. kernel code cannot accidentaly use an error in the Posix compatibility section).

I would advice against the term 'legacy code' for ported/external software. It sort of seems to imply that software is slated to be replaced by some 'non-legacy' (HelenOS native?) software in the future, which is not necessarily the case.

le-jzr commented:
@jxsvoboda Thanks for the comments.

> You shouldn't use double leading underscore in macro names / symbolic variables. That is the part of the namespace reserved for the compiler. Just use a single leading underscore (e.g. _errno_xx)

That's a bit of a misconception. Identifiers with two leading underscores, or one underscore and a capital letter, are reserved for the "language implementation", which is a fuzzy concept, but generally includes implementations of standard libraries and headers. A standards compliant program should be able to `#include <errno.h>` and use any non-reserved identifiers (including _errno_xx; especially _errno_xx, because single underscore prefix is generally reserved for file-level identifiers) without conflicting with anything in the header. So in short, double leading underscore is a standards-compliant solution. A single underscore isn't.

> If you are putting all the Exxx constants in a single header split into sections, you might want to make sure that code that is not supposed to make use of a particular section cannot do so by mistake (e.g. kernel code cannot accidentaly use an error in the Posix compatibility section).

Good idea. It shouldn't be very difficult.

> I would advice against the term 'legacy code' for ported/external software. It sort of seems to imply that software is slated to be replaced by some 'non-legacy' (HelenOS native?) software in the future, which is not necessarily the case.

You are right. I dislike this use of "legacy" myself, don't know why I wrote it like that. :/

le-jzr commented:
@jxsvoboda re: prefixes

It would make sense to use a prefix that's sure not to conflict with any future versions of any compiler, which might be something like `__helenos_xxx`. But in this case, I'd suggest that any existing instances of double-underscore prefix should be updated to this as well.

le-jzr commented:
To check the massive commit `e9dae33b4a0dd297560857151100e328c7f72484` in the middle:
```
git checkout e9dae33b4a0dd297560857151100e328c7f72484
git reset HEAD^
git add .
tools/srepl '\berrno_t\b' int
git add .
tools/srepl '\bsys_errno_t\b' sysarg_t
git add .
tools/srepl '(case_errno_t) ' ''
git reset
git diff
```

le-jzr commented:
I fixed issues for all configurations tested by `make check`, and addressed two of @jxsvoboda's suggestions. Going to test some harbor builds now, but I don't expect any problems there.

jermar commented on abi/include/abi/errno.h:
```diff
@@ -1,38 +1,117 @@
+ * software. These are not used in HelenOS code, but are defined here in
+ * order to avoid nasty hacks in libposix.
+ *
+ * If you decide to use one of these in native HelenOS code,
```
I think this split is not sustainable. People will simply start using those without moving them anywhere. Think of EPARTY, which was only supposed to be used by the kernel.

I also don't like that POSIX errnos are present in a HelenOS native header at all. It would be better if you could find a way to move these to libposix.

jermar commented on abi/include/abi/errno.in:
```diff
@@ -1,37 +1,116 @@
+ * software. These are not used in HelenOS code, but are defined here in
+ * order to avoid nasty hacks in libposix.
+ *
+ * If you decide to use one of these in native HelenOS code,
```
Same as the previous comment.

jermar commented on kernel/arch/amd64/include/arch/boot/boot.h:
```diff
@@ -44,7 +44,7 @@
 
 /* This is only a symbol so the type is dummy. Obtain the value using &. */
-extern int _hardcoded_unmapped_size;
+extern errno_t _hardcoded_unmapped_size;
```
Why errno_t here? As the comment says, we are interested only in the address...

jermar commented on kernel/arch/amd64/include/arch/cpuid.h:
```diff
@@ -56,7 +56,7 @@ typedef struct {
 } __attribute__ ((packed)) cpu_info_t;
 
-extern int has_cpuid(void);
+extern errno_t has_cpuid(void);
```
This is rather a true/false or no bits set / some bits set variable, there is no error information in it.

jermar replied:
Maybe an unsigned type would be best here.

le-jzr replied:
Maybe, but that's out of scope of this changeset. :)

jermar commented on kernel/arch/ia32/include/arch/boot/boot.h:
```diff
@@ -44,7 +44,7 @@
 
 /* This is only a symbol so the type is dummy. Obtain the value using &. */
-extern int _hardcoded_unmapped_size;
+extern errno_t _hardcoded_unmapped_size;
```
Again, there is no error information in this. We only care for the address of this symbol.

jermar commented on kernel/genarch/src/ddi/ddi-dummy.c:
```diff
@@ -47,7 +47,7 @@
-int ddi_iospace_enable_arch(task_t *task, uintptr_t ioaddr, size_t size)
+errno_t ddi_iospace_enable_arch(task_t *task, uintptr_t ioaddr, size_t size)
 {
 	return 0;
```
return EOK?

jermar commented on kernel/genarch/src/ddi/ddi-dummy.c:
```diff
@@ -62,7 +62,7 @@ int ddi_iospace_enable_arch(task_t *task, uintptr_t ioaddr, size_t size)
-int ddi_iospace_disable_arch(task_t *task, uintptr_t ioaddr, size_t size)
+errno_t ddi_iospace_disable_arch(task_t *task, uintptr_t ioaddr, size_t size)
 {
 	return 0;
```
return EOK?

jermar commented on kernel/generic/include/context.h:
```diff
@@ -44,7 +44,7 @@
 	} while (0)
 
-extern int context_save_arch(context_t *ctx) __attribute__((returns_twice));
+extern errno_t context_save_arch(context_t *ctx) __attribute__((returns_twice));
```
This is again rather true (the context save case) / false (the context restore case) case, I don't think context_save_arch can fail.

jermar commented on kernel/generic/src/cap/cap.c:
```diff
@@ -260,7 +260,7 @@ static bool cap_reclaimer(ht_link_t *link, void *arg)
  *
  * @return Negative error code in case of error.
```
Is it still negative?

jermar commented on kernel/generic/src/ddi/ddi.c:
```diff
@@ -343,29 +343,29 @@ NO_TRACE static int iospace_disable(task_id_t id, uintptr_t ioaddr, size_t size)
 	ddi_ioarg_t arg;
-	int rc = copy_from_uspace(&arg, uspace_io_arg, sizeof(ddi_ioarg_t));
+	errno_t rc = copy_from_uspace(&arg, uspace_io_arg, sizeof(ddi_ioarg_t));
 	if (rc != 0)
```
EOK?

jermar commented on kernel/generic/src/ddi/ddi.c:
```diff
@@ -343,29 +343,29 @@ NO_TRACE static int iospace_disable(task_id_t id, uintptr_t ioaddr, size_t size)
 	ddi_ioarg_t arg;
-	int rc = copy_from_uspace(&arg, uspace_io_arg, sizeof(ddi_ioarg_t));
+	errno_t rc = copy_from_uspace(&arg, uspace_io_arg, sizeof(ddi_ioarg_t));
 	if (rc != 0)
```
EOK?

jermar commented on kernel/generic/src/ipc/ipcrsc.c:
```diff
@@ -169,10 +169,10 @@ static kobject_ops_t phone_kobject_ops = {
  *
  * @return  Negative error code if a new capability cannot be allocated.
```
Drop Negative

jermar commented on uspace/app/download/main.c:
```diff
@@ -242,7 +242,7 @@ int main(int argc, char *argv[])
 	}
 
-	return EOK;
+	return 0;
```
This is travelling in the opposite direction.

le-jzr replied:
Huh. Weird.

le-jzr replied:
Ah, right, that's `main()`. Literal 0 is correct there, unless we make a convention of returning `EXXX` from `main()`.

le-jzr replied:
The story here is that I wrapped all errno returns from main in `EXIT_RC()` to get it to typecheck (simultaneously changing EOKs to 0 there), then I removed the macro in one of the last commits.

jermar commented on uspace/dist/src/c/demos/tetris/tetris.c:
```diff
@@ -68,17 +68,17 @@ static volatile const char copyright[] =
 char key_msg[100];
-int showpreview;
-int classic;
+errno_t showpreview;
```
This is actually a flag.

jermar commented on uspace/dist/src/c/demos/tetris/tetris.c:
```diff
@@ -68,17 +68,17 @@ static volatile const char copyright[] =
 
 long fallrate;
-int score;
+errno_t score;
```
An actual score, not an errno.

jermar commented on uspace/dist/src/c/demos/tetris/tetris.c:
```diff
@@ -68,17 +68,17 @@ static volatile const char copyright[] =
-int Rows;
-int Cols;
+errno_t Rows;
+errno_t Cols;
```
Rows and Cols are rows and cols, not errnos. Also, this is a copy of the actual file, so I guess the same issue exists in the original.

le-jzr replied:
Actually, the issue is caused by it being an unbuilt copy. I forgot to correct the "demos" because they are not compiled. In fact, it seems wrong to have copies like this in the repository. What's up the reason for not copying it in the build script? This way, you have to manually copy over any changes to the original.

jermar replied:
A good point actually, I don't know why :-)

jermar commented on uspace/drv/block/ahci/ahci.c:
```diff
@@ -108,22 +108,22 @@
 static void ahci_ahci_hw_start(ahci_dev_t *);
 
-static int ahci_dev_add(ddf_dev_t *);
+static errno_t ahci_dev_add(ddf_dev_t *);
```
I think there is a newly introduced trailing ws here

le-jzr replied:
It's not new.

jermar commented on uspace/lib/bithenge/src/failure.c:
```diff
@@ -61,9 +61,9 @@
-static int g_failure_index_selected = -1;
+static errno_t g_initialized = 0;
+static errno_t g_failure_index = -1;
+static errno_t g_failure_index_selected = -1;
```
These seem like non-errno variables to me.

le-jzr replied:
Hmm, yeah. Another file that is never built in HelenOS.

le-jzr replied:
I need to think of an easy way to list source files that haven't been compiled during the build.

jermar commented on uspace/lib/c/generic/futex.c:
```diff
@@ -49,7 +49,7 @@ void futex_initialize(futex_t *futex, int val)
 #ifdef FUTEX_UPGRADABLE
 
-int _upgrade_futexes = 0;
+errno_t _upgrade_futexes = 0;
```
I think this is actually a bool.

jermar commented on uspace/lib/c/test/adt/circ_buf.c:
```diff
@@ -37,7 +37,7 @@ enum {
 };
 
-static int buffer[buffer_size];
+static errno_t buffer[buffer_size];
```
Probably not what you wanted.

le-jzr commented:
I'm feeling very tired today. I'll leave this until tomorrow.

le-jzr commented:
I've decided to further split this into two separate merges. The first half will land in mainline in a few minutes.

le-jzr commented:
I'm closing this for now. The main problem (nonstandard error codes in libc) was resolved.
IMO we should eventually come up with a clean way to do the static checking (to make sure the self-documenting `errno_t` type is used where appropriate, as a matter of consistent style), but it's not a priority for the moment. In the meantime, it doesn't make much sense to keep this open.
